doc/design.md

```diff
@@ -0,0 +1,178 @@
+* relies on [JWT](https://jwt.io/) and [OAuth 2.0](https://oauth.net/2/)
+  authentication for inter-operability with other services
+* treats storage entirely separately from the API which purely handles data.
+  Several storage solutions can be supported, such as SSH for self-contained
+  systems, AzureFiles and S3 for Cloud deployments etc.
+
+A few things **aren't changing**:
+
+* [MongoDB](https://www.mongodb.com/) has been used with the first backend for
+  several years and is providing good results.  We might just move it to a
+  Cloud-based hosting such as Atlas for the future linux.kernelci.org
+  deployment.
+* [Redis]((https://redis.io/)) is still being used internally as in-memory
+  database for various features in addition to the pub/sub channels
+
+## Pipeline design
+
+Here's a simplified view of the current pipeline:
+
+```mermaid
+graph LR
+  trigger(Trigger) --> new{New <br /> kernel?}
+  new --> |yes| tarball(Tarball)
+  new -.-> |no| trigger
+  tarball --> |event: <br /> new revision| scheduler(Test <br /> Scheduler)
+  scheduler --> |event: <br /> test running| runtime(Runtime <br /> Environment)
+  runtime --> |event: <br /> test complete| email(Email <br /> Report)
+```
+
+Each step is a client for the API and related storage solutions.  They are all
+implemented in Python, rely on the [KernelCI Core](/docs/core) tools and run in
+a separate `docker-compose` container.  Here's a summary of each step:
+
+### Trigger
+
+The Trigger step periodically checks whether a new kernel revision has appeared
+on a number of git branches.  It first gets the revision checksum (SHA-1) from
+the top of the branch from the remote git repo and then queries the API to
+check whether it has already been covered.  If there's no revision entry with
+this checksum in the API, it then pushes a new `checkout` one with "running"
+state.
+
+### Tarball
+
+The Tarball step listens for pub/sub events about new revisions or `checkout`
+nodes.  When one is received, typically because the trigger pushed a new one to
+the API, it then updates a local git checkout of the full kernel source tree.
+Then it makes a tarball with the source code and pushes it to the storage.
+Finally, it sends an update to the `checkout` node to the API with the URL of
+the tarball added to the list of artifects and the status set to "available".
+
+### Scheduler
+
+The Scheduler step listens for a variety of pub/sub events and then looks for
+matches in the `scheduler` [YAML
+configuration](https://github.com/kernelci/kernelci-pipeline/blob/main/config/pipeline.yaml#L144).
+Every time an event matches an entry in the config, it will send a new node
+with the "running" state to the API and submit a job in the designated runtime
+environment.
+
+> **Note**: It's then up to the job itself to send the results to the API and
+> any files to storage.  The scheduler doesn't track the status of submitted
+> jobs, however the timeout steps will update the status of nodes that reach
```

Comment:
```suggestion
> jobs, however, the timeout step will update the status of nodes that reach
```

Reply:
No this is plural on purpose, there are several timeout services.  I guess I should rephrase it to make it clearer though.